Add Exponential Histogram

ocelotl opened the pull request:
Fixes #2963

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,316 @@ def collect(
+            low = buckets._index_start
+            high = index
+
+        if is_rescaling_needed:
```
Suggest breaking out the logic of this branch into a separate method, it's getting long...

ocelotl replied:
I first had it split as it is [here](https://github.com/lightstep/otel-launcher-go/blob/c9ca8483be067a39ab306b09060446e7fda65f35/lightstep/sdk/metric/aggregator/histogram/structure/exponential.go#L361) but it was hard reading it since `incrementIndexBy` actually may increment the index or not.

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,316 @@ def collect(
+                change += 1
+
+            if change == 0:
+                return
```
This is too soon to return, I think--the update below would be skipped.
OTOH, I think this is unreachable code? I think you can raise an exception because of an internal error, rescaling is not needed when `change == 0`.

ocelotl replied:
Right, that return should not be there, removed. And right that is unreachable code, I guess it can also be removed from [here](https://github.com/lightstep/otel-launcher-go/blob/c9ca8483be067a39ab306b09060446e7fda65f35/lightstep/sdk/metric/aggregator/histogram/structure/exponential.go#L330) as well :v:

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,316 @@ def collect(
+        self._instrument_temporality = AggregationTemporality.DELTA
+        self._start_time_unix_nano = start_time_unix_nano
+
+    def aggregate(self, measurement: Measurement) -> None:
```
```suggestion
    def aggregate(self, measurement: Measurement) -> None:
        # Outline of this method:
        # 1. Maintain the _min, _max, _count, _zero_count, and _sum fields
        # 2. Determine the range (positive or negative), then use the absolute value
        # 3. Compute the (absolute) value's index at the current scale
        # 4. Check whether the value's index falls into [_index_start, _index_end]
        # 5. If not, rescale the data structure with a change of scale
        # 6. Then, if the new index falls outside [_index_start, _index_end], adjust the boundaries of the circular backing array
        # 7. Compute the position of the index in the circular backing array
        # 8. Finally, update the bucket.
```

Reading through this, I wanted for comments. (Possibly they should go in-line with the code.)

ocelotl replied:
Nice! added comments.

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,316 @@ def collect(
+            max_ = self._max
+            min_ = self._min
+
+            self._negative._counts = [0]
+            self._positive._counts = [0]
```
Should the _index_start, _index_end, _index_base fields be updated to match this reset?

ocelotl replied:
Right, fixed.

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,316 @@ def collect(
+            attributes=self._attributes,
+            start_time_unix_nano=start_time_unix_nano,
+            time_unix_nano=collection_start_nano,
+            count=(
+                sum(self._negative.counts)
+                + sum(self._positive.counts)
+                + self._zero_count
+            ),
```
I believe this equals `_count` (from line 439), you shouldn't need to sum the buckets at this point.

ocelotl replied:
Fixed

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,316 @@ def collect(
+            ),
+            # FIXME: Find the right value for flags
+            flags=0,
+            min=min_,
+            max=max_,
```
I suppose the exporter has to handle the case where `count_ == 0` meaning that min and max are +/- infinity. As long as the exporter omits these, it's OK. Otherwise, since the protocol has these as optional, should these be None when `count_ == 0`?

ocelotl replied:
Right, fixed :v:

jmacd commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/buckets.py:
```diff
@@ -0,0 +1,148 @@
+
+        needed = needed - 1
+
+        needed |= needed >> 1
+        needed |= needed >> 2
+        needed |= needed >> 4
+        needed |= needed >> 8
+        needed |= needed >> 16
```
I can't remember what this does? 🤣  (j/k but please add a comment)

ocelotl replied:
Right, reimplemented and added a comment.

ocelotl commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/metric_exporter/__init__.py:
```diff
@@ -266,6 +267,40 @@ def _translate_data(
+                                negative=pb2.ExponentialHistogramDataPoint.Buckets(
+                                    offset=data_point.negative.offset(),
+                                    bucket_counts=data_point.negative.bucket_counts,
+                                ),
```
Fixed

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+            raise Exception("size {max_size} is smaller than {min_max_size}")
+
+        if max_size > self._max_max_size:
+            raise Exception("size {max_size} is larter than {max_max_size}")
```
```suggestion
            raise Exception(f"size {max_size} is larger than {max_max_size}")
```

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+        super().__init__(attributes)
+        # max_size is the maximum capacity of the positive and negative
+        # buckets.
+        if max_size < self._min_max_size:
+            raise Exception("size {max_size} is smaller than {min_max_size}")
+
+        if max_size > self._max_max_size:
+            raise Exception("size {max_size} is larter than {max_max_size}")
```
These error messages should be more descriptive about what invariant is being violated, "size 0 is less than 2" is not very descriptive.

I would also recommend throwing [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError)

ocelotl replied:
Right, fixed.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+        # positive and negative buckets.
+        self._mapping = LogarithmMapping(LogarithmMapping._max_scale)
+
+        self._instrument_temporality = AggregationTemporality.DELTA
```
Why is this not coming from the constructor like for other aggregators?

ocelotl replied:
We do the same for the `_ExplicitBucketHistogramAggregation`. In fact, only `_SumAggregation` has `instrument_temporality` as a constructor parameter.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+                index > buckets._index_end
+                and (index - buckets._index_start) >= self._max_size
+            ):
+                raise Exception("Downscaling logic error")
```
Can you make a more descriptive message and ideally specific exception type?

Should we be throwing here at all?

ocelotl replied:
Right, I removed it.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+        """
+
+        with self._lock:
+            if not any(self._negative.counts) and not any(
+                self._positive.counts
+            ):
```
would this not work?
```suggestion
            if self._count == 0:
```

ocelotl replied:
Makes sense :+1:

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+            if min_ == inf:
+                min_ = None
+
+        if self._count == self._zero_count:
+            scale = 0
```
why does this change the scale?

ocelotl replied:
Scale does not matter if all values are zero.

aabmass replied:
Should we use zero (protobuf default values) in these cases or `None`? I think some places we do `None`

ocelotl replied:
Good question, @jmacd what do you think?

jmacd replied:
I don't have an opinion on `None` vs `0`. I would expect the protobuf field is not actually encoded because it is a scalar and has the default value.

The reason this may not matter much is that when the structure transitions from 0 measurements to 1 measurement, the scale is supposed to reset to the maximum scale again, so the empty state is less important: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#use-the-maximum-scale-for-single-measurements

ocelotl replied:
All right, I think we can leave it as `0` then. :+1:

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+            if min_ == inf:
+                min_ = None
+
+        if self._count == self._zero_count:
```
Why don't we need to continue holding the lock here?

ocelotl replied:
Good catch, fixed.

aabmass reviewed (changes requested):
I think there are some thread safety issues in this PR, can you review some of the classes missing locks or where locks are not held and non-atomic operations happen?

ocelotl commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,334 @@ def collect(
+        # _count is the count of all calls to aggregate.
+        self._count = 0
+
+        # _zero_count is the count of all the calls to aggregate when the value
```
> Doesn't this include values near zero but not exactly to zero (something about subnormals?)

I think it does: `zero_count` is kept to populate [this](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L523), described as:

>   // `zero_count` is the count of values that are either exactly zero or
  // within the region considered zero by the instrumentation at the
  // tolerated degree of precision.  This bucket stores values that
  // cannot be expressed using the standard exponential formula as
  // well as values that have been rounded to zero.

The smallest possible float point number that can be represented with IEEE 754 is a _denormal_ number, [this one](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/ieee_754.md#normal-and-denormal-floating-point-numbers):

$$ 5 \times 10^{-324} $$

Assuming that the limit for _"the tolerated degree of precision for the instrumentation"_ (as mentioned above) is the smallest denormal value, then just comparing the value with 0 as we do [here](https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R453) should work, because Python treats values smaller than $5 \times 10^{-324}$ to be equal to $5 \times 10^{-324}$ or equal to zero:

``` python
Python 3.9.15 (main, Nov 28 2022, 20:14:06) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.7.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: 5e-324 == 0
Out[1]: False

In [2]: 5e-325 == 0
Out[2]: True

In [3]: 4e-325
Out[3]: 0.0

In [4]: 4e-324
Out[4]: 5e-324

In [5]: 3e-324
Out[5]: 5e-324

In [6]: 2e-324
Out[6]: 0.0

In [7]: 1e-324
Out[7]: 0.0

In [8]: 1e-324 == 0
Out[8]: True

In [9]: 4e-324 == 5e-324
Out[9]: True
```

ocelotl replied:
> https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#zero-count-and-zero-threshold
> 
> do we need to implement this? it looks like we are just checking if it is exactly zero

Right, we could implement `zero_threshold` too. Since that is an optional parameter we need a default behavior, I think the above one works for our purposes.

aabmass replied:
> Assuming that the limit for "the tolerated degree of precision for the instrumentation" (as mentioned above) is the smallest denormal value

The mapping functions appear to map all subnormal values to a single index:

https://github.com/open-telemetry/opentelemetry-python/blob/edf4d6cd07b5ea922180b52c9e737b7be0d59f9f/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/exponent_mapping.py#L90-L91

https://github.com/open-telemetry/opentelemetry-python/blob/edf4d6cd07b5ea922180b52c9e737b7be0d59f9f/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/logarithm_mapping.py#L103-L105

Is that bucket not considered the zero count since all values less than it (including zero) are mapped to that index?

ocelotl replied:
> Is that bucket not considered the zero count since all values less than it (including zero) are mapped to that index?

Not including zero, `map_to_index` is not called if the value to aggregate is zero:

https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R455

aabmass replied:
Based on SIG discussion, I'm going to resolve this for now.

ocelotl commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -433,6 +774,27 @@ def _create_aggregation(
         raise Exception(f"Invalid instrument type {type(instrument)} found")
 
 
+class ExponentialBucketHistogramAggregateion(Aggregation):
```
good catch, fixed

ocelotl commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/instrument.py:
```diff
@@ -212,6 +212,28 @@ def record(
         )
 
 
+class ExponentialHistogram(_Synchronous, APIHistogram):
```
I have no idea :sweat_smile: Removed.

srikanthccv commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,329 @@ def collect(
+                )
+            ]
+            positive_counts = [
+                curr_count + prev_count
```
Shouldn't this be `curr_count - prev_count `

ocelotl replied:
Correct, fixed

srikanthccv commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/buckets.py:
```diff
@@ -0,0 +1,149 @@
+            self._counts = self._counts[::-1]
+            # [4, 3, 2, 1, 0]
+
+            self._counts = self._counts[:bias][::-1] + self._counts[bias:]
+            # [3, 4, 2, 1, 0]
+
+            self._counts = self._counts[:bias] + self._counts[bias:][::-1]
+            # [3, 4, 0, 1, 2] This is a rotation of the backing array.
```
```suggestion
            self._counts = self._counts[:bias][::-1] + self._counts[bias:][::-1]
            # [3, 4, 0, 1, 2] This is a rotation of the backing array.
```

ocelotl commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,329 @@ def collect(
+                change += 1
+
+            if change < 0:
+                raise Exception(f"Impossible change of scale: {change}")
```
Right, removed

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,326 @@ def collect(
+
+        if value == 0:
+            self._zero_count += 1
+            return
```
Can you just add a comment explaining why we return here?

ocelotl replied:
Done

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,326 @@ def collect(
+        self._instrument_temporality = AggregationTemporality.DELTA
+        self._start_time_unix_nano = start_time_unix_nano
+
+    def aggregate(self, measurement: Measurement) -> None:
```
I don't think this method is thread safe

ocelotl replied:
Added a lock for the whole method.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,326 @@ def collect(
+        # 7. Increment the bucket.
+        buckets.increment_bucket(bucket_index)
+
+    def collect(
```
This doesn't look thread safe. IIRC, nothing is preventing new calls to `aggregate()` from coming in during collection.

It looks like we may have the same issues in ExplicitBucketHistogramAggregation and SumAggregation, maybe im misremembering if we actually do hold a lock in the code that calls this?

ocelotl replied:
Added a lock for the whole method.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/buckets.py:
```diff
@@ -0,0 +1,148 @@
+
+        # Index of the 0th position in self._counts: self._counts[0] is the
+        # count in the bucket with index self._index_base.
+        self._index_base = 0
+
+        # self._index_start is the smallest index value represented in
+        # self._counts.
+        self._index_start = 0
+
+        # self._index_start is the largest index value represented in
+        # self._counts.
+        self._index_end = 0
```
It looks like the `_ExponentialBucketHistogramAggregation` is reading/setting these fields. Since it's already in `_internal`, I think we can should make these public properties to make it clear that these are being modified from outside of this class

jsuereth commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,326 @@ def collect(
+
+        self._previous_point = current_point
+
+        with self._lock:
```
It seems like all this "self.X" stuff should be ABOVE in the same first lock.

I'd recommend copying all the data to local variables *or* keeping the lock on this entire method here if you want it to be accurate.

I thought that's how you'd done the other aggregations, but as @aabmass called out, this may not be the case and we need to reevaluate the other ones too.

ocelotl replied:
Added the lock for the whole method.

jsuereth commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,326 @@ def collect(
+
+    # _min_max_size is is the smallest reasonable value which is small enough
+    # to contain the entire normal floating point range at the minimum scale.
+    _min_max_size = 2
```
Is this limit part of the specification?

ocelotl replied:
It does not need to be, it is a mathematical consequence of representing floating point numbers in IEEE 754 format. I added a [test case](https://github.com/open-telemetry/opentelemetry-python/pull/2964/commits/fccfccbd5c4cdba78ec5a60d6c57bb698653889e) that show how this happens.

jmacd replied:
I see this as a minimum viable size for a histogram to be useful. A size=1 or size=0 histogram is not sensible. With minimum scale of -10, we get one bucket for (0x1p-1024, 1], and one bucket for (1, 0x1p1024]. So I see this as a consequence of the minimum scale being set correctly.

jsuereth commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,326 @@ def collect(
+
+            change = 0
+
+            while high - low >= self._max_size:
```
Nit: This "new scale" algorithm is dying to be called out as a helper method.  Something like:

```python
def get_scale_reduction(start, end, current):
```

defined directly on `Buckets`.

Right now, this code is harder to follow and would really benefit from some abstraction to simplify responsibility.  There's a lot of code here "reaching into" the Buckets class.

ocelotl replied:
That rescaling algorithm is used in our code only in one place, I prefer not to have "single-use functions" because that makes the code hard to read, the reader has to make a mental note of the values passed to the function, scroll up or down to find the function (or find it in another file), paste the values, read the function code and come back to the place where the function call was.

ocelotl replied:
Hm, I think I was confused by the line mentioned above, I see now how we are setting private buckets values like `_index_start` and `_index_end` in this code, it may be possible to make a new function for the `Buckets` class that does this, as you mention :+1:

ocelotl replied:
After taking another look, I have refactored this by adding properties for the buckets attributes accessed here. I prefer not to create a function for this because it is very hard to find a descriptive name for it.

jsuereth reviewed (commented):
There seems to be some threading issues, also recommend some general cleanup for easier reading of this code.  It's tricky so ideally it should be as easy to read as possible.  I think more helper methods  (or classes + responsibilities) could help there.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/buckets.py:
```diff
@@ -0,0 +1,172 @@
+from math import ceil, log2
+
+
+class Buckets:
```
Does this class need to be thread safe? If not it would be good to leave a comment for the reader

ocelotl replied:
Added a comment.

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,330 @@ def collect(
+        if max_size > self._max_max_size:
+            raise ValueError(
+                f"Buckets max size {max_size} is larger than "
+                "maximum max size {self._max_max_size}"
```
```suggestion
                f"maximum max size {self._max_max_size}"
```

aabmass commented on opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py:
```diff
@@ -349,6 +362,330 @@ def collect(
+        if max_size < self._min_max_size:
+            raise ValueError(
+                f"Buckets max size {max_size} is smaller than "
+                "minimum max size {self._min_max_size}"
```
```suggestion
                f"minimum max size {self._min_max_size}"
```

srikanthccv reviewed (commented):
@ocelotl where is scale difference b/w previous and current points are handled?

jmacd reviewed (approved):
👏

ocelotl commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/metric_exporter/__init__.py:
```diff
@@ -266,6 +267,51 @@ def _translate_data(
+                                attributes=self._translate_attributes(
+                                    data_point.attributes
+                                ),
+                                time_unix_nano=data_point.time_unix_nano,
+                                start_time_unix_nano=(
+                                    data_point.start_time_unix_nano
+                                ),
```
That's right, it is a Python thing to keep lines under [80 characters](https://peps.python.org/pep-0008/#maximum-line-length).

ocelotl commented on exporter/opentelemetry-exporter-otlp-proto-grpc/src/opentelemetry/exporter/otlp/proto/grpc/metric_exporter/__init__.py:
```diff
@@ -60,6 +60,7 @@
     ResourceMetrics,
     ScopeMetrics,
     Sum,
+    ExponentialHistogram as ExponentialHistogramType,
```
Seems like it is [already](https://github.com/open-telemetry/opentelemetry-python/blob/main/exporter/opentelemetry-exporter-otlp-proto-grpc/pyproject.toml#L34).

srikanthccv reviewed (changes requested):
Working with @ocelotl on the collection part, requesting changes to prevent the accidental merge.

ocelotl commented:
Work in progress to address the issues pointed out by @srikanthccv, moving to draft for the time being.

ocelotl commented:
@srikanthccv PTAL, I added merging methods to `collect`.

srikanthccv commented:
force push is making it really hard to follow your recent changes addressing the comment.

ocelotl commented:
> force push is making it really hard to follow your recent changes addressing the comment.

Weird, I must have done so accidentally, pretty much this is what is being added here:
https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R633-R654
https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R664-R677
https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R685-R698

We now use call a `_merge` method that uses rescales the histogram based on the previous point.

srikanthccv commented:
You force-pushed again, ah? Can you help me with the changes you made the since last time?
